### PR TITLE
feat(#213): add terrain mesh export and example

### LINKED ISSUE #213: [shiplog/plan] Heightmap terrain plugin with DEM tile loader
## Motivation

Many Galeon projects need to render terrain from authored heightmap data: a terrain mesh + sample queries are a generic engine primitive. Today the engine treats terrain as the application's problem — there is no first-class `Terrain` resource, no heightmap-to-mesh pipeline, and no example showing how to wire a heightmap mesh into the Three.js sync layer.

This issue adds a thin **`engine-terrain`** plugin that ingests an authored heightmap (PNG16 or 32-bit float), produces a Three.js plane-geometry mesh with vertex normals, and exposes a `Terrain` ECS resource with `height_at(x, z)`, `normal_at(x, z)`, and bounding-box accessors.

Real-world DEM (Copernicus / SRTM / national-mapping-agency) ingestion is a separate concern that builds on top of this primitive — tracked downstream rather than in core engine.

## Out of scope

- DEM / GeoTIFF ingestion (downstream)
- Tile streaming / LOD (followup issue)
- Splat-mapping, vegetation, biome painting
- Physics integration (point queries are enough for now)

## Tasks

- [ ] **T1: Define `Terrain` resource and `HeightmapPlugin`** `[tier-2]`
  - **What:** New crate `crates/engine-terrain` with `Terrain` resource (origin, size, pixel-stride, sample fn) and a `HeightmapPlugin` that registers it.
  - **Verification:** `cargo test -p engine-terrain` — `height_at` returns expected values for a synthetic 4x4 heightmap.

- [ ] **T2: PNG16 loader** `[tier-2]`
  - **What:** Read PNG16 grayscale into `Vec<f32>` heights, with min/max metadata and optional vertical exaggeration.
  - **Verification:** Unit test loads a checked-in 16x16 PNG16 fixture, asserts corner values.

- [ ] **T3: Mesh generation + `engine-three-sync` integration** `[tier-3]`
  - **What:** Generate plane-geometry vertex buffer from heights; emit as a renderable entity through the existing `FramePacket` channel. Vertex normals from finite differences.
  - **Verification:** Headless extraction test confirms vertex count = (W+1)*(H+1) and one expected normal direction.

- [ ] **T4: Example: `examples/terrain`** `[tier-2]`
  - **What:** Minimal app that loads a synthetic ridge heightmap (procedurally generated PNG16) and renders it via three-sync. Demonstrates `height_at(x,z)` and `normal_at(x,z)` queries against the loaded `Terrain` resource.
  - **Verification:** `cargo run --example terrain` produces a frame containing terrain mesh entity in the FramePacket; sample queries return expected values within tolerance.

Authored-by: claude/opus-4.7 (claude-code)
Updated-by: claude/opus-4.7 (claude-code)
Edit-kind: amendment
Edit-note: Narrowed scope to engine primitive only — removed DEM / GeoTIFF / national-mapping-agency ingestion (now downstream concern). task_count 5→4, max_tier unchanged.

### PR DESCRIPTION
**Pull request by devallibus**
## Summary

This PR completes #213 by adding the terrain mesh/export half of the heightmap primitive. It keeps Terrain as the authoritative Rust CPU resource, adds a generated TerrainMesh adapter buffer, and emits an opt-in renderable terrain entity through the existing three-sync MeshHandle/MaterialHandle frame-packet path.

Closes #213

## Review Status

- **Current state:** awaiting re-review after feedback fix
- **Last reviewed by:** chatgpt-codex-connector
- **Last reviewed at:** 2026-05-02T16:12:42Z
- **Reviewed commit:** 4ba580f
- **Source artifact:** https://github.com/galeon-engine/galeon/pull/237#pullrequestreview-4215232864
- **Needs re-review since:** 2111142

### Initial Plan
Finish the remaining #213 tasks after PR #231 delivered the Terrain resource and PNG16 loader: T3 mesh generation / three-sync integration and T4 native terrain example.

### What We Discovered
- Current FramePacket carries renderable entity rows and asset handles, not vertex payloads. The implementation therefore generates a Rust-side TerrainMesh buffer and uses existing Transform + MeshHandle + MaterialHandle extraction rather than widening the packet contract.
- A Bevy prior-art refresh confirmed the data-resource vs render-adapter split and the need to pin triangle winding / +Y normals in docs and tests.

### Implementation Issues
- None blocking. The only adjustment was preserving the current frame-packet contract instead of introducing mesh payload transport in this issue.

### Key Decisions Made

| Decision | Choice | Why |
|----------|--------|-----|
| Terrain data ownership | Keep Terrain authoritative and make TerrainMesh an adapter | Matches Bevy prior art and Galeon's Rust-first rule |
| Frame transport | Emit a render entity with mesh/material handles | Avoids expanding FramePacket for vertex data before the renderer asset contract exists |
| Mesh convention | Row-major source samples, local X/Z positions, +Y winding/normals | Gives Three.js consumers a stable geometry contract |

### Changes Made

**Commits:**
- 4ba580f feat(#213/T3-T4): add terrain mesh export and example
- 2111142 fix(#213): compute terrain mesh normals from grid samples

## Testing

- [x] cargo fmt --all -- --check
- [x] cargo test -p galeon-engine-terrain
- [x] cargo run -p galeon-engine-terrain --example terrain
- [x] cargo clippy -p galeon-engine-terrain --all-targets -- -D warnings
- [x] cargo check --workspace
- [x] cargo run --example terrain
- [x] cargo test --workspace
- [x] git diff --check

**Verification summary:** Added terrain mesh unit coverage, a three-sync extraction integration assertion from the terrain crate, and a native example smoke that loads procedural PNG16 terrain, generates a mesh, samples height/normal, and extracts a terrain render entity. Review follow-up adds an edge-normal regression for a 0.1-sized 4x4 grid and verifies mesh normals are generated from integer source samples.

## Stacked PRs / Related

- PR #231 delivered #213 T1/T2.
- PR #235 and PR #236 landed related follow-up cleanup from PR #231 review.

## Knowledge for Future Reference

TerrainMesh is intentionally a CPU-side Galeon buffer, not a Three.js type. TS consumers still need a small adapter that converts positions/normals/indices into THREE.BufferGeometry and registers it against the chosen mesh handle. UVs, splat maps, collider export, Sobel normals, and tiled LOD remain out of scope.

---
Authored-by: openai/gpt-5 (codex)
Updated-by: openai/gpt-5 (codex)
Edit-kind: amendment
Edit-note: Added review feedback commit, re-review status, and edge-normal regression verification.
Last-code-by: openai/gpt-5 (codex)
*Captain's log - PR timeline by [shiplog](https://github.com/devallibus/shiplog)*

### CODE REVIEW

**Comment by devallibus** [comment]
## [shiplog/review-handoff] #213: PR #237 cross-model review request

**Status:** awaiting independent review

**Scope:** Full diff of PR #237, commit 4ba580f.

**Review contract:**
- Verify TerrainMesh generation preserves row-major X/Z sample mapping, upward-facing winding, and normal expectations.
- Verify HeightmapPlugin::with_render_mesh does not widen the FramePacket contract and emits a normal renderable entity through MeshHandle / MaterialHandle extraction.
- Verify the example and docs honestly describe the current boundary: TS consumers still convert TerrainMesh into THREE.BufferGeometry themselves.
- Confirm the verification commands in the PR body are sufficient or request targeted additions.

**Cross-model gate:** This code was authored by openai/gpt-5 (codex), so closure needs review by a different model identity before merge.

Authored-by: openai/gpt-5 (codex)

File: crates/engine-terrain/src/lib.rs [hunk 1]
```diff
@@ -249,6 +251,90 @@ impl Terrain {
...
+            for x in 0..width {
+                let local_x = x as f32 * stride_x;
+                let local_z = z as f32 * stride_z;
+                positions.extend_from_slice(&[local_x, terrain.sample_at(x, z), local_z]);
+
+                let world_x = terrain.origin[0] + local_x;
+                let world_z = terrain.origin[1] + local_z;
+                let normal = terrain
+                    .normal_at(world_x, world_z)
+                    .unwrap_or([0.0, 1.0, 0.0]);
```
[the remark this answers is not in the document]
Comment (devallibus) [reply crates/engine-terrain/src/lib.rs]: Fixed in 2111142.

`TerrainMesh::from_terrain` now computes normals from integer source grid coordinates via `normal_at_sample`, so edge vertices no longer route through world-space `normal_at` and cannot fall back to `[0, 1, 0]` because of `origin + size` rounding. I also added `terrain_mesh_computes_edge_normals_from_grid_samples` for the `size=0.1`, `width=4` case.

Verification: `cargo fmt --all -- --check`; `cargo test -p galeon-engine-terrain`; `cargo clippy -p galeon-engine-terrain --all-targets -- -D warnings`; `cargo check --workspace`; `git diff --check`.

**Comment by devallibus** [comment]
@codex review
